openslides_backend/action/actions/user/save_saml_account.py

```diff
@@ -0,0 +1,158 @@
+from ....shared.interfaces.event import Event, EventType
+from ....shared.patterns import fqid_from_collection_and_id
+from ....shared.schema import schema_version
+from ....shared.util import ONE_ORGANIZATION_ID
+from ...generics.create import CreateAction
+from ...generics.update import UpdateAction
+from ...mixins.send_email_mixin import EmailCheckMixin
+from ...mixins.singular_action_mixin import SingularActionMixin
+from ...util.action_type import ActionType
+from ...util.default_schema import DefaultSchema
+from ...util.register import register_action
+from ...util.typing import ActionData, ActionResultElement
+from .user_mixin import UsernameMixin
+
+allowed_user_fields = [
+    "saml_id",
+    "title",
+    "first_name",
+    "last_name",
+    "email",
+    "gender",
+    "pronoun",
+    "is_active",
+    "is_physical_person",
+]
+
+
+@register_action("user.save_saml_account", action_type=ActionType.STACK_INTERNAL)
+class UserSaveSamlAccount(
+    EmailCheckMixin, UsernameMixin, CreateAction, SingularActionMixin
+):
+    """
+    Internal action to save (create or update) a saml account.
+    It should be called from the auth service.
+    """
+
+    user: Dict[str, Any] = {}
+    saml_attr_mapping: Dict[str, str]
+    check_email_field = "email"
+    model = User()
+    schema = DefaultSchema(User()).get_default_schema()
+    skip_archived_meeting_check = True
+
+    def validate_instance(self, instance: Dict[str, Any]) -> None:
+        organization = self.datastore.get(
+            ONE_ORGANIZATION_FQID,
+            ["saml_enabled", "saml_attr_mapping"],
+            lock_result=False,
+        )
+        if not organization.get("saml_enabled"):
+            raise ActionException(
+                "SingleSignOn is not enabled in OpenSlides configuration"
+            )
+        self.saml_attr_mapping = organization.get("saml_attr_mapping", {})
+        if not self.saml_attr_mapping or not isinstance(self.saml_attr_mapping, dict):
+            raise ActionException(
+                "SingleSignOn field attributes are not configured in OpenSlides"
+            )
+        additional_required_fields = {
+            value: self.model.saml_id.get_payload_schema()
+            for key, value in self.saml_attr_mapping.items()
+            if key == "saml_id"
+        }
+        additional_optional_fields = {
+            value: cast(Field, getattr(self.model, key, {})).get_payload_schema()
+            for key, value in self.saml_attr_mapping.items()
+            if key != "saml_id" and key in allowed_user_fields
+        }
+        self.schema = {
+            "$schema": schema_version,
+            "title": "create saml account schema",
+            "type": "object",
+            "properties": {
+                **additional_required_fields,
+                **additional_optional_fields,
+            },
+            "required": list(additional_required_fields.keys()),
+            "additionalProperties": True,
+        }
```

Comment:
Multiple remarks:
- You do not need to split the fields artificially into required and optional, just to merge them together again afterwards
- You can use `model.get_field()` instead of `getattr()` to get the field
- You do not need to check for the `allowed_user_fields` since it is guaranteed from `organization.update` that these are correct
- I would use more verbose names instead of `key` and `value`
```suggestion
        schema = {
            "$schema": schema_version,
            "title": "create saml account schema",
            "type": "object",
            "properties": {
                payload_field: self.model.get_field(model_field).get_payload_schema()
                for model_field, payload_field in self.saml_attr_mapping.items()
            },
            "required": self.saml_attr_mapping["saml_id"],
            "additionalProperties": True,
        }
```

Reply:
> Multiple remarks:
> 
>     * You do not need to split the fields artificially into required and optional, just to merge them together again afterwards
done
> 
>     * You can use `model.get_field()` instead of `getattr()` to get the field
> 
done
>     * You do not need to check for the `allowed_user_fields` since it is guaranteed from `organization.update` that these are correct
> 
no, because this could be changed

>     * I would use more verbose names instead of `key` and `value`

done